### DISCUSSION TITLE
Question on use of SetFixedCostOfAllVehicles for PDP problems

### DISCUSSION THREAD

**Discussion by mmflynn** [google/or-tools#3011] [Routing (and legacy CP) questions]
I am using OR Tools for small pickup and delivery problems, allocating an estimate of the number of vehicles required.  In small problems I can see that more vehicles are run than required resulting in a greater overall distance travelled (although when I query the objective function I see this extra distance and extra vehicles results in a lower objective value ??).

Using the PDP example, I get,
|Vehicles |Distance|Objective|
|---------|---------|----------|
|3 |5980m.|229580|
|4 | 6916m| 226116|
|5| 7760m|185760|
|6 | 7760m |185760 *| 

*one vehicle is left unused.

When I add SetFixedCostOfAllVehicles(1000), I have the same pattern, except the distances in the solution printer are increased by 1000 (which appears to be the fixed cost being added).

This is all based on the example code for the PDP problem.  I am looking for what is driving the Objective (as it is clearly not simply distance minimisation) and why SetFixedCostOfAllVehicles does not incentivise a reduction in the use of vehicles.

Thanks.

**Comment by mmflynn** [comment]
I have found part of the answer.

In the sample code where the distance dimension and PDP sequencing are set
```python
    # Add Distance constraint.
    dimension_name = 'Distance'
    routing.AddDimension(
        transit_callback_index,
        0,  # no slack
        3000,  # vehicle maximum travel distance
        True,  # start cumul to zero
        dimension_name)
    distance_dimension = routing.GetDimensionOrDie(dimension_name)
    distance_dimension.SetGlobalSpanCostCoefficient(100)

    # Define Transportation Requests.
    for request in data['pickups_deliveries']:
        pickup_index = manager.NodeToIndex(request[0])
        delivery_index = manager.NodeToIndex(request[1])
        routing.AddPickupAndDelivery(pickup_index, delivery_index)
        routing.solver().Add(
            routing.VehicleVar(pickup_index) == routing.VehicleVar(
                delivery_index))
        routing.solver().Add(
            distance_dimension.CumulVar(pickup_index) <=
            distance_dimension.CumulVar(delivery_index)
```

At the end of establishing the Distance dimension a GlobalSpanCostCoeffficient is set.  This appears to equalise the work across the available vehicles via an additional cost in the objective function

`distance_dimension.SetGlobalSpanCostCoefficient(100)`

If this line is removed, the results are

Vehicles | Distance | Objective|Vehicles not used|
-- | -- | --|--
3 | 6460m. | 6460|-
4 | 6460m | 64606|1
5 | 6460m | 6460|2
6 | 6460m | 6460 |3

This clearly performs better at selecting the fleet to optimise distance, but without the SpanCostCoefficient, we reach a consistently inferior solution in terms of sequencing stops.  Adding back in SetGlobalSpanCostCoefficient(5) and I get the expected behaviour of 3 vehicles with a solution of 5980m.  The objective is 10452, which obviously is the distance plus some manipulation of the SpanCostCoefficient.   It would be great if someone could help me understand what SpanCostCoefficient is actually doing here!.

**Comment by jmarca** [comment]
It is minimizing the difference between the smallest route and the largest route for the dimension in question. By minimizing that largest difference, all routes are pulled towards a similar value. 

The c++ source code is enlightening

James
